docs: ADR Self Monitor Storage (#1436)
kyma-project/telemetry-manager: docs/contributor/arch/014-telemetry-self-monitor-storage.md

The current storage configuration and retention policy for the self-monitoring data are not well-defined. Currently, some installations face the issue that self-monitoring storage fills up and exceeds the storage limit despite the retention policies of 2 hours or 50 MBytes. 
The Telemetry self-monitoring data is stored in the Prometheus TSDB, which is designed for large-scale deployments. The amount of data collected by the Telemetry self-monitoring is actually small compared to the Prometheus capabilities (a few MBytes). Nevertheless, the storage size and retention policies must be carefully configured.

### Storage and Retention with TSDB

The TSDB storage size-based retention works as follows: It includes data blocks like the write-ahead-log (WAL), the checkpoints, the m-mapped chunks, and the persistent blocks. The TSDB counts all those data blocks to decide performing any retention.

> Even if size of all those data blocks go beyond the configured retention size, only persistence blocks are deleted because the WAL, checkpoints, and m-mapped chunks that are required for normal operation of TSDB. The WAL segments can grow up to 128MB before compacting, and Prometheus will keep at least 3 WAL files; so-called 2/3 rules. To ensure that Telemetry self-monitoring doesn't exceed the storage limit, minimum storage volume size should be calculated to be at least 3 * WAL segment size + some more space for other data types.  

Reviewer: ```suggestion
Even if the size of all those data blocks exceeds the configured retention size, only persistence blocks are deleted because the WAL, checkpoints, and m-mapped chunks are needed for normal operation of TSDB. The WAL segments can grow up to 128MB before compacting, and Prometheus will keep at least 3 WAL files; so-called 2/3 rules. To ensure that Telemetry self-monitoring doesn't exceed the storage limit, minimum storage volume size should be calculated to be at least 3 * WAL segment size + some more space for the other data types.  
```

> # 14. Telemetry Self Monitoring Storage

Reviewer: ```suggestion
# 14. Telemetry Self-Monitoring Storage
```
You use a hyphen everywhere else, so I'd add it here too.

Strictly speaking, you need a hyphen when it's an adjective and no hyphen (just blank space) when it's a compound noun ("self monitoring is crucial...") , but I fear that might confuse readers.

Date: 2024-10-09

## Status

Proposed

## Context

The Telemetry module self-monitoring is crucial for the overall health of the system. The self-monitoring data is used to detect issues in the Telemetry module and to provide insights into the system's health. The self-monitoring data is stored in a time-series database (TSDB) and is used to generate alerts. 

> The current storage configuration and retention policy for the self-monitoring data are not well-defined, currently, some installation faces the issue self-monitoring storage fill-up and exceed the storage limit despite the retention policies 2 hours or 50 MBytes. 

Reviewer: ```suggestion
The current storage configuration and retention policy for the self-monitoring data are not well-defined. Currently, some installations face the issue that self-monitoring storage fills up and exceeds the storage limit despite the retention policies of 2 hours or 50 MBytes. 
```
```suggestion
The current storage configuration and retention policy for the self-monitoring data are not well-defined, currently, some installation faces the issue self-monitoring storage fill-up and exceed the storage limit despite the retention policies 2 hours or 50 MBytes. 
```

> The Telemetry self-monitoring data is stored in the Prometheus TSDB, which is designed for large scale deployments, the amount data collected by the Telemetry self-monitoring is actually small compared to the Prometheus capabilities (currently few MBytes) nevertheless the storage size and retention policies have to be carefully configured.

Reviewer: ```suggestion
The Telemetry self-monitoring data is stored in the Prometheus TSDB, which is designed for large-scale deployments. The amount of data collected by the Telemetry self-monitoring is actually small compared to the Prometheus capabilities (a few MBytes). Nevertheless, the storage size and retention policies must be carefully configured.
```

### Storage and Retention with TSDB

> The TSDB storage size-based retention works in a way, it includes data blocks the WAL, checkpoints, m-mapped chunks, and persistent blocks. The TSDB although counts all of those storage blocks to decide any deletion, the WAL, checkpoints, and m-mapped chunks required for normal operation of TSDB.

Reviewer: - _"The TSDB storage size-based retention works in a way, it includes data blocks the WAL, checkpoints, m-mapped chunks, and persistent blocks."_
This grammar is confusing, especially this bit _"...works in a way, it includes data blocks the WAL,..."_. Is blocks a noun ("data blocks") or a verb ("it ... blocks the WAL")? Please explain what this sentnce means, or rephrase it and/or add more punctuation to clarify it. 
- Similarly, what does _"decide any deletion, the WAL, checkpoints, and m-mapped chunks"_ mean? 

- Furthermore, we should explain that WAL means "write-ahead log" the first time we mention it.

> Only persistence blocks are deleted even if all those data blocks go beyond the configured retention size. The WAL segments can grow up to 128MB before compacting, and Prometheus will keep at least 3 WAL files so called 2/3 rules. To ensure the Telemetry self-monitoring doesn't exceed the storage limit, minimum storage volume size should be calculated at least 3 * WAL segment size + some more space for other data types.  

Reviewer: not sure what "so called 2/3 rules" means.  Just guessing in the following suggestion:

```suggestion
Even if all those data blocks go beyond the configured retention size, only persistence blocks are deleted. The WAL segments can grow up to 128MB before compacting, and Prometheus will keep at least 3 WAL files; so-called 2/3 rules. To ensure that Telemetry self-monitoring doesn't exceed the storage limit, minimum storage volume size should be calculated to be at least 3 * WAL segment size + some more space for other data types.  
```

Do we need the words "so-called 2/3 rules" at all? Which value do they add?

### TSDB Storage architecture and retention

> For more information and insight into the Prometheus storage architecture and retention policy, please refer to the [Prometheus TSDB: Compaction and Retention](https://ganeshvernekar.com/blog/prometheus-tsdb-compaction-and-retention).

Reviewer: ```suggestion
For more information about the Prometheus storage architecture and retention policy, see [Prometheus TSDB: Compaction and Retention](https://ganeshvernekar.com/blog/prometheus-tsdb-compaction-and-retention).
```

> The TSDB WAL and checkpoint architecture is described in the [Prometheus TSDB: WAL and Checkpoint](https://ganeshvernekar.com/blog/prometheus-tsdb-wal-and-checkpoint/).

Reviewer: ```suggestion
For the TSDB WAL and checkpoint architecture, see [Prometheus TSDB: WAL and Checkpoint](https://ganeshvernekar.com/blog/prometheus-tsdb-wal-and-checkpoint/).
```

## Consequences

> The Telemetry self-monitoring requires or collect very small amount of data for operation (currently few MBytes), despite the small amount of data, the storage size have to be at least 500MByte for a normal and safe operation.

Reviewer: ```suggestion
Even though the Telemetry self-monitoring collects very little data for operation (currently, a few MBytes), the storage size must be at least 500MByte for a normal and safe operation.
```
